File: 백준/Silver/1874. 스택 수열/스택 수열.js
const [n, ...input] = require("fs")
  .readFileSync(0)
  .toString()
  .split("\n")
  .map(Number);

const stack = [];
let result = [];
let count = 1;

for (let i = 0; i < n; i += 1) {
  while (count <= input[i]) {
    stack.push(count);
    result.push("+");
    count += 1;
  }
  if (stack.pop() !== input[i]) {
    result = ["NO"];
    break;
  }
  result.push("-");
}
console.log(result.join("\n"));

File: 백준/Silver/1874. 스택 수열/스택 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Deque;

public class Main {

  public static void main(String[] args) throws IOException {
    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    int N = Integer.parseInt(br.readLine());

    StringBuilder sb = new StringBuilder();
    Deque<Integer> stack = new ArrayDeque<>();

    int count=1;
    for(int i = 0; i < N; i++) {
      int number = Integer.parseInt(br.readLine());
      while(stack.isEmpty() || number > stack.peekLast()){
        stack.offerLast(count);
        count++;
        sb.append("+\n");
      }

      if(stack.peekLast()==number){
        sb.append("-\n");
        stack.pollLast();
      }
    }
    if(stack.isEmpty()){
      System.out.println(sb);
    }else{
      System.out.println("NO");
    }
  }
}
